types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,274 @@
+	}
+}
+
+void wlr_pointer_gesture_swipe_notify_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wl_resource *swipe;
+	wl_resource_for_each(swipe, &gestures->swipes) {
+		struct wl_client *swipe_client = wl_resource_get_client(swipe);
+		if (swipe_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(swipe_client));
+		zwp_pointer_gesture_swipe_v1_send_end(swipe, serial,
+				time_msec, cancelled);
+	}
+}
+
+static const struct zwp_pointer_gesture_swipe_v1_interface swipe_impl = {
+	.destroy = resource_handle_destroy,
+};
+
+static void get_swipe_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(gestures_resource);
+
+	struct wl_resource *swipe = wl_resource_create(client,
+		&zwp_pointer_gesture_swipe_v1_interface,
+		wl_resource_get_version(gestures_resource),
+		id);
+	if (swipe == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(swipe, &swipe_impl, NULL, resource_remove_from_list);
+	wl_list_insert(&gestures->swipes, wl_resource_get_link(swipe));
+}
+
+void wlr_pointer_gesture_pinch_notify_begin(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		uint32_t fingers) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(pinch_client));
+		zwp_pointer_gesture_pinch_v1_send_begin(pinch, serial,
+				time_msec, focus->resource, fingers);
+	}
+}
+
+void wlr_pointer_gesture_pinch_notify_update(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		double dx,
+		double dy,
+		double scale,
+		double rotation) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
+			continue;
+		}
+		zwp_pointer_gesture_pinch_v1_send_update(pinch, time_msec,
+				wl_fixed_from_double(dx), wl_fixed_from_double(dy),
+				wl_fixed_from_double(scale),
+				wl_fixed_from_double(rotation));
+	}
+}
+
+void wlr_pointer_gesture_pinch_notify_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_surface *focus,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wl_resource *pinch;
+	wl_resource_for_each(pinch, &gestures->pinches) {
+		struct wl_client *pinch_client = wl_resource_get_client(pinch);
+		if (pinch_client != wl_resource_get_client(focus->resource)) {
```

Comment:
This doesn't respect seats. The `wl_pointer` object used to create the pinch object is bound to a seat.